qualtran/surface_code/fifteen_to_one.py

```diff
@@ -0,0 +1,380 @@
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+from functools import lru_cache
+
+import cirq
+import numpy as np
+from attrs import frozen
+
+from qualtran.surface_code.magic_count import MagicCount
+from qualtran.surface_code.magic_state_factory import MagicStateFactory
+from qualtran.surface_code.quantum_error_correction_scheme_summary import (
+    FowlerSuperconductingQubits,
+    QuantumErrorCorrectionSchemeSummary,
+)
+from qualtran.surface_code.t_factory_utils import PauliRotationChannel, storage_error
+
+
+@frozen
+class FifteenToOne(MagicStateFactory):
+    """15-to-1 Magic T state factory.
+
+        reference: https://arxiv.org/abs/1905.06903
+
+    Attributes:
+        d_X: Side length of the surface code along which X measurements happen.
+        d_Z: Side length of the surface code along which Z measurements happen.
+        d_m: Number of code cycles used in lattice surgery.
+        qec: Quantum error correction scheme being used.
+    """
+
+    d_X: int
+    d_Z: int
+    d_m: int
+    qec: QuantumErrorCorrectionSchemeSummary = FowlerSuperconductingQubits
+
+    def footprint(self) -> int:
+        return 2 * (self.d_X + 4 * self.d_Z) * 3 * self.d_X + 4 * self.d_m
+
+    @lru_cache(5)
+    def _final_state(self, phys_err: float):
+        factory = _build_factory(phys_err, self.d_X, self.d_Z, self.d_m, self.qec)
+        return (
+            cirq.DensityMatrixSimulator(dtype=np.complex128).simulate(factory).final_density_matrix
+        )
+
+    @lru_cache(5)
+    def p_fail(self, phys_err: float) -> float:
+        projector = np.kron(np.eye(2), np.ones((16, 16)) / 16)
+        return np.real(1 - np.trace(projector @ self._final_state(phys_err)))
+
+    @lru_cache(5)
+    def p_out(self, phys_err: float) -> float:
+        projector = np.kron(
+            np.eye(2), np.ones((16, 16)) / 16
+        )  # I \otimes ones \otimes ones \otimes ones \otimes ones / 16
+        project_state = (
+            1
+            / (1 - self.p_fail(phys_err))
+            * (projector @ self._final_state(phys_err) @ projector.T.conj())
+        )
+        T_state = np.array([1, np.exp(-1j * np.pi / 4)]).reshape((1, 2)) / np.sqrt(
+            2
+        )  # |T><T| \otimes ones \otimes ones \otimes ones \otimes ones / 16
+        target_density = np.kron(T_state.T.conj() @ T_state, np.ones((16, 16)) / 16)
+        return np.real(1 - np.trace(project_state @ target_density))
+
+    def n_cycles(self, n_magic: MagicCount, phys_err: float) -> int:
+        """The number of cycles (time) required to produce the requested number of magic states."""
+        num_t = n_magic.n_t + 4 * n_magic.n_ccz
+        return np.ceil(num_t * 6 * self.d_m / (1 - self.p_fail(phys_err)))
```

Comment:
a reference for the equation for the factor of `6` would be appreciated here.

Why are you dividing by the success probability? Is this supposed to take into account the number of cycles "wasted" by a bad distillation? Do other factories take this into account or do they assume everything was fine and you have to look at the failure probability to figure out this correction? The failure probability includes both heralded and unheralded errors: if you get a bad distillation and have to try again you increase the number of cycles but don't increase the error of your algorithm. If you have a data error during the distillation you don't notice, you don't increase the number of cycles, but you do increase the error rate of your actual algorithm.

Reply:
added reference. yes I think this formula is the expected nubmer of attempts until accepting a state 

$$
E[\mathrm{attempts}] = \sum_{i = 1}^\infty i \times p_\mathrm{fail}^{i-1} (1 - p_\mathrm{fail}) = \frac{1}{1 - p_\mathrm{fail}}
$$

each attempt has $6 d_m$ code cycles.

---

> if you get a bad distillation and have to try again

yes this is what is caught here. the number of tries until we meeting the acceptance criteria

> If you have a data error during the distillation you don't notice, you don't increase the number of cycles, but you do increase the error rate of your actual algorithm.

That is `p_out` which I use in computing the `distillation_error`

---

from what I understand from your comment it seems $6 d_m$ would be more in line with the only other factory we have (the CCZ factory). which way do we wanna go from here?

Reply:
I don't know what the best way forward is. Maybe just add a comment that this includes both sources of increased time and the other factory doesn't